### PR TITLE
Integration execution group to fluent API

### PR DESCRIPTION
**Pull request by junchuanwang**
- Integrated the execution group into the template
- The FluentClient interface has been renamed to ParSeqBasedFluentClient

### CODE REVIEW

File: restli-tools/src/main/resources/apiVmTemplates/action.vm [hunk 1]
```diff
@@ -15,138 +15,127 @@ Copyright (c) 2021 LinkedIn Corp.
...
-    public CompletionStage<${method.valueClassDisplayName}> ${util.nameCamelCase(${method.name})}(
-      #if(${method.isEntityAction()})
-      $spec.keyClassName $spec.idName #if(${method.hasActionParams()}),#end
-      #end
-      #if(${method.hasActionParams()})
-      Function<$actionParamClassName, $actionParamClassName> paramsProvider
-      #end
-    );
+      #if(${method.hasRequiredParams()}) ## action with no required Params will only have one API
+        #if($method.hasOptionalParams()) ## when have optional params, the optionalParamsProvider is optional
```
Comment (junchuanwang) [inline restli-tools/src/main/resources/apiVmTemplates/action.vm:24]: I added this so it is now consistent with other apis: There is an additional API that does not need the optionalParam as argument

File: restli-tools/src/main/resources/macros/library.vm [hunk 2]
```diff
@@ -177,21 +177,49 @@
...
   #foreach($param in $method.requiredParameters)
-    $param.paramClassDisplayName $param.paramName#if( $foreach.hasNext || $hasOptionalParams),#end
+    $param.paramClassDisplayName $param.paramName#if( $foreach.hasNext || ($hasOptionalParams || $withEG)),#end
   #end
   #if($hasOptionalParams)
-    Function<${util.restMethodToClassPrefix($method.methodName)}OptionalParameters, ${util.restMethodToClassPrefix($method.methodName)}OptionalParameters> optionalParamsProvider
+    Function<${util.restMethodToClassPrefix($method.methodName)}OptionalParameters, ${util.restMethodToClassPrefix($method.methodName)}OptionalParameters> optionalParamsProvider #if($withEG),#end
+  #end 
+  #if($withEG)ExecutionGroup executionGroup #end
+
```
Comment (karthikbalasub) [inline restli-tools/src/main/resources/macros/library.vm:196]: remove this line

File: restli-int-test/src/test/java/com/linkedin/restli/examples/TestParSeqBasedFluentClientApiWithExecutionGroup.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    results.clear();
+    ExecutionGroup eg = greetings.generateExecutionGroup();
+    results.add(greetings.get(1L, eg));
+    results.add(greetings.get(1L, eg));
+    results.add(greetings.get(1L, eg));
+    results.add(greetings.get(1L, eg));
+    assert(results.get(0) instanceof ParSeqBasedCompletionStage);
+    t = ((ParSeqBasedCompletionStage<Greeting>) results.get(0)).getTask();
+    Assert.assertFalse(t.isDone());
+    Assert.assertFalse(hasTask("greetings batch_get(reqs: 1, ids: 3)",t.getTrace()));
```
[the remark this answers is not in the document]
Comment (junchuanwang) [reply restli-int-test/src/test/java/com/linkedin/restli/examples/TestParSeqBasedFluentClientApiWithExecutionGroup.java:125]: I am looking for a more intuitive way. Do you have a better way other than checking the number of "get" calls in the Trace?

File: restli-int-test/src/test/java/com/linkedin/restli/examples/TestParSeqBasedFluentClientApiWithExecutionGroup.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  /**
+    This test is to test if the batch method has been triggered.
+
+    Note: Currently ParSeqRestClient only support "batch" for gets.
+  */
+  @Test public void testBatchGet() throws Exception
+  {
+    // Test 3+1 = 4 calls using the "runBatchOnClient" method in the fluent Client
```
[the remark this answers is not in the document]
Comment (karthikbalasub) [reply restli-int-test/src/test/java/com/linkedin/restli/examples/TestParSeqBasedFluentClientApiWithExecutionGroup.java:89]: Please update the comment to refer the constants.

File: restli-int-test/src/test/java/com/linkedin/restli/examples/TestParSeqBasedFluentClientApiWithExecutionGroup.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (karthikbalasub) [reply restli-int-test/src/test/java/com/linkedin/restli/examples/TestParSeqBasedFluentClientApiWithExecutionGroup.java]: Did you check how the parseq batching tests verify correctness?

[the remark this answers is not in the document]
Comment (junchuanwang) [reply restli-int-test/src/test/java/com/linkedin/restli/examples/TestParSeqBasedFluentClientApiWithExecutionGroup.java]: That uses trace (same as what I was doing right now), but that is "checking" trace, somewhat similar to checking the log, less ideal for testing.

**Pull request merged**